### PR TITLE
Refactor/remove cast constructors

### LINKED ISSUE #64: Remove "cast" constructors
Right now, there are constructors in most types that take an argument of the same type,

For example,
```kt
class PointerType {
    public constructor(type: Type): this(type.ref)
}
```
This is deeply problematic because of this

```kt

val voidPointer = PointerType(VoidType()) // this is a void instead of void*. Yikes

```

This caused my languages test suites to segfault on generated code. I suggest removing these constructors and do explicit casting using the old way .asSomeType() and so on

### PR DESCRIPTION
**Pull request by dhruvrajvanshi**
Fixes #64 and adds a few APIs

### CODE REVIEW

File: src/main/kotlin/dev/supergrecko/kllvm/core/typedefs/Builder.kt [hunk 1]
```diff
@@ -106,6 +108,27 @@ public class Builder public constructor(context: Context = Context.getGlobalCont
...
+        return InstructionValue(LLVM.LLVMBuildStore(ref, value.ref, toPointer.ref))
+    }
+
+    public fun buildExtractValue(aggVal: Value, index: Int, name: String): InstructionValue {
+        return InstructionValue(LLVM.LLVMBuildExtractValue(ref, aggVal.ref, index, name))
+    }
+
+    public fun buildStructGEP(pointer: Value, index: Int, name: String): InstructionValue {
+        return InstructionValue(LLVM.LLVMBuildStructGEP(ref, pointer.ref, index, name))
+    }
```
Comment (junlarsen) [inline src/main/kotlin/dev/supergrecko/kllvm/core/typedefs/Builder.kt:130]: I think we want to create a nicer API for the builder in the future but this is good enough for now.

**Review by junlarsen — APPROVED** [review]

**Comment by junlarsen** [comment]
Seems like the tests need an update after the constructors have been removed. Could you check the "Allow edits from maintainers" box so I can push the test changes I've made?

**Comment by dhruvrajvanshi** [comment]
Sorry my bad. I couldn't find the button to allow edits but it was a small change so I've pushed the fix myself.

**Comment by dhruvrajvanshi** [comment]
Okay now actually fixed 😅 for real

**Comment by dhruvrajvanshi** [comment]
Also adding ref constructor validations for types so that we can throw early exceptions instead of segfaults

**Pull request merged**